py/dml/c_backend.py

```diff
@@ -2623,6 +2814,173 @@ def register_saved_attributes(initcode, node):
                  postindent = -1)
     initcode.out('}\n', preindent=-1)
     initcode.out('}\n', preindent=-1)
+
+def generate_hook_attr_names(node, prefix):
+    # Generate attributes for all subcomponents.
+    for child in sorted(node.get_components(),
+                        key=lambda o: o.name or ''):
+        if child.objtype == 'hook':
+            yield (child, prefix + child.ident)
+        elif (isinstance(child, objects.CompositeObject)
+              and child.objtype not in {'bank', 'port', 'subdevice'}):
+            yield from generate_hook_attr_names(
+                child, prefix + param_str(child, 'name') + "_")
+
+def register_hook_attributes(initcode, node):
+    if node is dml.globals.device:
+        for port in node.get_components('bank', 'port', 'subdevice'):
+            register_hook_attributes(initcode, port)
+        cls = 'class'
+        getter = '_DML_get_hook_attr'
+        setter = '_DML_set_hook_attr'
+    else:
+        if dml.globals.dml_version == (1, 2) and not node.name:
+            # anonymous bank
+            return
+        cls = port_class_ident(node)
+        getter = '_DML_get_port_hook_attr'
+        setter = '_DML_set_port_hook_attr'
+
+    hook_attr_names = list(generate_hook_attr_names(node, ''))
+    if not hook_attr_names:
+        return
+
+    initcode.out('{\n', postindent=1)
+    initcode.out('const char *hook_attr_names[%d] = {\n'
+                 % (len(hook_attr_names),), postindent=1)
+    for (hook, attr_name) in hook_attr_names:
+        register_attribute(
+            hook.site, None if node is dml.globals.device else node, attr_name)
+        initcode.out(f'"{attr_name}",\n')
+    initcode.out('};\n', preindent=-1)
+    initcode.out("const uint32 hook_ids[%d] = {\n" % (len(hook_attr_names),),
+                 postindent=1)
+    for (hook, _) in hook_attr_names:
+        initcode.out(f'{hook.uniq},\n')
+    initcode.out('};\n', preindent=-1)
+    initcode.out('for (uint32 idx = 0; idx < %d; ++idx) {\n'
+                 % (len(hook_attr_names),) , postindent = 1)
+    initcode.out(f'_DML_register_hook_attribute({cls}, hook_attr_names[idx], '
+                 + f'{getter}, {setter}, &_hook_aux_infos[hook_ids[idx] - 1], '
+                 + f'_hook_id_infos[hook_ids[idx] - 1], {node.dimensions});\n')
+    initcode.out('}\n', preindent=-1)
+    initcode.out('}\n', preindent=-1)
+
+
+def generate_hook_attribute_funs():
```

Comment:
Significant bodies of verbatim C code, maybe it's time to pay the cost (in terms of boilerplate in ctree-test, I suppose?) of moving these into dmllib.h. Can be pushed to separate PR.

I suppose you put the functions here because they reference global static arrays. One option could be optional forward declarations, like:

dmllib.h:
```
#ifndef SKIP_STATIC_ARRAYS
static ht_str_table_t _typeseq_after_on_hook_hts[];
attr_value_t
_DML_get_hook_attr(...)
...
#endif
```
and `-DSKIP_STATIC_ARRAYS` when compiling ctree-test

Reply:
I'm gonna do this in a separate PR, or perhaps a future review round when all else is settled -- this will require some thought and testing to ensure it Doesn't Go Wrong.